File: src/main/java/org/gga/graph/Edge.java
package org.gga.graph;

/**
 * @author mike
 */
public class Edge {
    private final int v;
    private final int w;
    private int idx;

    public Edge(int v, int w) {
        this.v = v;
        this.w = w;
    }

    public int v() {
        return v;
    }

    public int w() {
        return w;
    }

    public int idx() {
        return idx;
    }

    public void setIdx(int idx) {
        this.idx = idx;
    }

    public int other(final int t) {
        return from(t) ? w : v;
    }

    public boolean from(int t) {
        return t == v;
    }

    public String toString() {
        return "(" + v + " -> " + w + ")";
    }

    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Edge edge = (Edge) o;
        return v == edge.v && w == edge.w;

    }

    public int hashCode() {
        return v * 31 + w;
    }
}

File: src/main/scala/org/gga/graph/Edge.scala
package org.gga.graph

/**
 * Edge from v to w.
 *
 * @author mike
 */
class Edge(aFrom: Int, aTo: Int) {
  val v: Int = aFrom
  val w: Int = aTo
  private var idx: Int = 0

  def target = w
  def source = v

  def getIdx: Int = idx

  def setIdx(idx: Int) {
    this.idx = idx
  }

  def other(t: Int): Int = if (from(t)) w else v

  def from(t: Int): Boolean = t == v

  override def toString: String = "(" + v + " -> " + w + ")"

  override def equals(other: Any): Boolean = {
    other match {
      case that: Edge => v == that.v && w == that.w
      case _ => false
    }
  }

  override def hashCode: Int = v * 31 + w
}